File: backend/app/schemas/analysis.py
from typing import Any, Literal

from pydantic import BaseModel, Field

AnomalyClassification = Literal[
    "statistical_outlier",
    "logical_violation",
    "missing_data",
    "duplicate",
    "referential_inconsistency",
    "cross_column_inconsistency",
    "temporal_anomaly",
    "potential_business_anomaly",
    "confirmed_data_quality_issue",
    "unknown",
]
ConfidenceLevel = Literal["high", "medium", "low"]


class Finding(BaseModel):
    claim: str
    value: float | int | str | None = None
    unit: str | None = None
    affected_rows: int | None = None
    total_rows: int | None = None
    calculation: str | None = None
    classification: AnomalyClassification | None = None
    confidence: ConfidenceLevel = "medium"
    evidence: list[str] = Field(default_factory=list)
    # Set to False by app.ai.answer_validator when a finding's numbers
    # couldn't be matched to anything in the actual tool-call evidence —
    # never silently trusted just because the model said it.
    verified: bool = True


class ToolCallRecord(BaseModel):
    tool: str
    arguments: dict[str, Any]
    result: dict[str, Any]


class ConversationTurn(BaseModel):
    question: str
    answer: str


class AnalyzeRequest(BaseModel):
    question: str = Field(min_length=1, max_length=2000)
    conversation_history: list[ConversationTurn] = Field(default_factory=list, max_length=10)


class AnalyzeResponse(BaseModel):
    question: str
    answer: str
    findings: list[Finding] = Field(default_factory=list)
    tool_calls: list[ToolCallRecord] = Field(default_factory=list)
    needs_clarification: str | None = None
    status: Literal["ok", "degraded"] = "ok"
    warnings: list[str] = Field(default_factory=list)

File: frontend/src/types/analysis.ts
// Mirrors backend/app/schemas/analysis.py exactly — the hybrid AI
// Analyst's response shape (POST /datasets/{id}/analyze). Distinct from
// AskResponse (types/insight.ts), which is the older, single-query /ask
// path's shape.

export type AnomalyClassification =
  | 'statistical_outlier'
  | 'logical_violation'
  | 'missing_data'
  | 'duplicate'
  | 'referential_inconsistency'
  | 'cross_column_inconsistency'
  | 'temporal_anomaly'
  | 'potential_business_anomaly'
  | 'confirmed_data_quality_issue'
  | 'unknown'

export type ConfidenceLevel = 'high' | 'medium' | 'low'

export interface Finding {
  claim: string
  value: number | string | null
  unit: string | null
  affected_rows: number | null
  total_rows: number | null
  calculation: string | null
  classification: AnomalyClassification | null
  confidence: ConfidenceLevel
  evidence: string[]
  // Set by the backend's answer validator — false means this finding's
  // numbers couldn't be matched to anything in the actual tool-call
  // evidence. Never hidden when false; always shown, flagged.
  verified: boolean
}

export interface ToolCallRecord {
  tool: string
  arguments: Record<string, unknown>
  result: Record<string, unknown>
}

export interface ConversationTurn {
  question: string
  answer: string
}

export interface AnalyzeRequest {
  question: string
  conversation_history?: ConversationTurn[]
}

export interface AnalyzeResponse {
  question: string
  answer: string
  findings: Finding[]
  tool_calls: ToolCallRecord[]
  needs_clarification: string | null
  status: 'ok' | 'degraded'
  warnings: string[]
}

/** A tool_calls[].result carrying an "error" key is how app.ai.tool_specs
 * reports a tool failure — never an exception, always this shape. */
export function toolCallFailed(call: ToolCallRecord): string | null {
  const error = call.result?.error
  return typeof error === 'string' ? error : null
}
